p2p/pex/addrbook_test.go

```diff
@@ -589,6 +589,53 @@ func TestMultipleAddrBookAddressSelection(t *testing.T) {
 	}
 }
 
+func TestAddrBookAddDoesNotOverwriteOldIP(t *testing.T) {
+	fname := createTempFileName("addrbook_test")
+	defer deleteTempFile(fname)
+
+	// This test creates adds a peer to the address book and marks it good
+	// It then attempts to override the peer's IP, by adding a peer with the same ID
+	// but different IP. We distinguish the IP's by "RealIP" and "OverrideAttemptIP"
+	peerID := "678503e6c8f50db7279c7da3cb9b072aac4bc0d5"
+	peerRealIP := "1.1.1.1:26656"
+	peerOverrideAttemptIP := "2.2.2.2:26656"
+	SrcAddr := "b0dd378c3fbc4c156cd6d302a799f0d2e4227201@159.89.121.174:26656"
+
+	// There is a chance that AddAddress will ignore the new peer its given.
+	// So we repeat trying to override the peer several times,
+	// to ensure we aren't in a case that got probabilistically ignored
+	numOverrideAttempts := 10
+
+	peerRealAddr, err := p2p.NewNetAddressString(peerID + "@" + peerRealIP)
+	require.Nil(t, err)
+
+	peerOverrideAttemptAddr, err := p2p.NewNetAddressString(peerID + "@" + peerOverrideAttemptIP)
+	require.Nil(t, err)
+
+	src, err := p2p.NewNetAddressString(SrcAddr)
+	require.Nil(t, err)
+
+	book := NewAddrBook(fname, true)
+	book.SetLogger(log.TestingLogger())
+	err = book.AddAddress(peerRealAddr, src)
+	require.Nil(t, err)
+	book.MarkAttempt(peerRealAddr)
+	book.MarkGood(peerRealAddr.ID)
+
+	// Double check that adding a peer again doesn't error
+	err = book.AddAddress(peerRealAddr, src)
+	require.Nil(t, err)
+
+	// Try changing ip but keeping the same node id. (change 1.1.1.1 to 2.2.2.2)
+	// This should just be ignored, and not error.
```

Comment:
Maybe sending you peer's with the same ID but different IP should be deemed as malicious behaviour / worth lowering your peer score for?

Its tricky since you can't tell which peer sent you the correct IP.

Figured I should flag this in case its worth further discussion in a separate issue. (I don't think it blocks this, since we're still maintaining the old functionality atm)

Reply:
This sounds more suited to be discussed outside of this PR, maybe open up an issue with the open question?

Reply:
Will do!

Reply:
I feel like there may have been some initial motivation to allow the IP address associated with an ID to change. One could imaging rebooting a node eg on AWS or otherwise and keeping the same config data (ie. ID) but getting a new IP. Would such a case still be accounted for? Would such a peer now be treated as malicious?

Reply:
This is definitely really tricky to handle correctly. (I've been having trouble writing the issue, due to me having unclear thoughts on any particular solution idea)

In the code as of this PR, AddAddress would ignore the new node with a different IP. I think this means it would only send messages to the old IP. However noone is marked malicious. 

If we wanted to support this, we'd have to authenticate the incoming connection via secret connection before updating the IP, and maybe have an update methon on the address book.